lib/pf/UnifiedApi/Controller/Config/Roles.pm

```diff
@@ -26,7 +25,62 @@ has 'primary_key' => 'role_id';
 
 use pf::ConfigStore::Roles;
 use pfappserver::Form::Config::Roles;
-
+use pfconfig::cached_hash;
+
+tie our %RolesReverseLookup, 'pfconfig::cached_hash', 'resource::RolesReverseLookup';
+
+sub can_delete {
+    my ($self) = @_;
+    my ($status, $msg) = $self->can_delete_from_db();
+    if (is_error($status)) {
+        return ($status, $msg);
+    }
+
+    ($status, $msg) = $self->can_delete_from_config();
+    if (is_error($status)) {
+        return ($status, $msg);
+    }
+
+    return (200, '');
+}
+
+sub can_delete_from_config {
+    my ($self) = @_;
+    if (exists $RolesReverseLookup{$self->id}) {
+        return (422, 'Role still in use');
+    }
+
+    return (200, '');
+}
+
+my $CAN_DELETE_FROM_DB_SQL = <<SQL;
+SELECT count(*) from node_category
+WHERE
+    name = ? AND
+    (
+        EXISTS (SELECT mac from node WHERE node.category_id = node_category.category_id) OR
+        EXISTS (SELECT mac from node WHERE node.bypass_role_id = node_category.category_id) OR
+        EXISTS (SELECT security_event_id from `class` WHERE `class`.target_category = node_category.name) OR
+        EXISTS (SELECT category from password WHERE password.category = node_category.category_id)
+    );
+SQL
+
+sub can_delete_from_db {
+    my ($self) = @_;
+    my $id = $self->id;
+    my ($status, $sth) = pf::dal::node_category->db_execute($CAN_DELETE_FROM_DB_SQL, $id);
+    if (is_error($status)) {
+        return ($status, "Unable to check role in the database");
+    }
+
+    my ($count) = $sth->fetchrow_array();
+    $sth->finish;
+    if ($count) {
```

Comment:
From what I understand the code does (correct me if I'm wrong), this prevents a user from deleting a role if its still referenced in the DB.

What we need according to the discussions we had for this feature is to tell the user that its still used and allow him to:
- Assign a new role to the nodes that use this role prior to deletion
- Assign a new bypass role to the nodes that use this bypass role prior to deletion
- Assign a new role to the local users that use this role prior to deletion

These 3 should be done automatically for the user during the deletion without him needing to go track down all of these DB entries and modify them.

The security events check can stay as a check as the role would have to be changed in the security events config